src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp

```diff
@@ -0,0 +1,236 @@
+ *
+ * This implements the upwind multipenalty boundary correction term
+ * \f$D_\alpha\f$. The general form is given by:
+ *
+ * \f{align*}{
+ *   \label{eq:pnpm upwind boundary term characteristics}
+ *   D_\beta =
+ *   T_{\beta\hat{\beta}}^{\mathrm{ext}}
+ *   \Lambda^{\mathrm{ext},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{ext}}_{\hat{\alpha}}
+ *   -T_{\beta\hat{\beta}}^{\mathrm{int}}
+ *   \Lambda^{\mathrm{int},-}_{\hat{\beta}\hat{\alpha}}
+ *   v^{\mathrm{int}}_{\hat{\alpha}},
+ * \f}
+ *
+ * Note that Eq. (6.3) of \cite Teukolsky2015ega is not exactly what's
+ * implemented since that boundary term does not consistently treat both sides
+ * of the interface on the same footing.
+ *
+ * For the first-order GH system the correction is:
+ *
+ * \f{align}{
+ *   \label{eq:pnpm upwind gh metric}
+ *   D_{g_{ab}} &= \tilde{\lambda}_{v^{g}}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},g}_{ab}
+ *                - \tilde{\lambda}_{v^{g}}^{\mathrm{int}}
+ *                v^{\mathrm{int},g}_{ab}, \\
+ *   \label{eq:pnpm upwind gh pi}
+ *   D_{\Pi_{ab}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}_{ab} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}_{ab}\right)
+ *                + \tilde{\lambda}_{v^g}^\mathrm{ext}\gamma_2
+ *                v^{\mathrm{ext},g}_{ab}
+ *                \notag \\
+ *              &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)
+ *                - \tilde{\lambda}_{v^g}^\mathrm{int}\gamma_2
+ *                v^{\mathrm{int},g}_{ab} , \\
+ *   \label{eq:pnpm upwind gh phi}
+ *   D_{\Phi_{iab}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}_{ab}\right)n_i^{\mathrm{ext}}
+ *                + \tilde{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{iab}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{iab},
+ * \f}
+ *
+ * with characteristic fields
+ *
+ * \f{align}{
+ *   \label{eq:pnpm gh v g}
+ *   v^{g}_{ab} &= g_{ab}, \\
+ *   \label{eq:pnpm gh v 0}
+ *   v^{0}_{iab} &= (\delta^k_i-n^k n_i)\Phi_{kab}, \\
+ *   \label{eq:pnpm gh v pm}
+ *   v^{\pm}_{ab} &= \Pi_{ab}\pm n^i\Phi_{iab} -\gamma_2 g_{ab},
+ * \f}
+ *
+ * and characteristic speeds
+ *
+ * \f{align}{
+ *   \lambda_{v^g} =& -(1+\gamma_1)\beta^i n_i -v^i_g n_i, \\
```

Comment:
I'm curious: it looks like the mesh velocity term is just a correction to the shift, `\beta^i \rightarrow \beta^i + v^i_g`, except for this line, where you only apply the correction to the term without the constraint damping.

I'm just double-checking whether this is correct, and if so, if there is an easy way to understand why there's no term `-\gamma_1 v^i_g` here.

Reply:
Thanks! I think I get it now. You're subtracting a diagonal matrix from B^i_{\alpha\beta}, so that changes the eigenvalues but not the eigenvectors. Since it turns out that the char speeds except one all involve \beta^I n_i once, subtracting v^i_g n_i acts like changing the shift, but not quite, because of that one exception, the gamma_1 term.

Reply:
Is it right to think that the shift `N^i` is changed by `-v^i_g` everywhere it appears on its own contracting a `d_i u_\alpha`, except at this one place when `u_\alpha = \psi_{ab}` where it appears with a prefactor as `(1 + \gamma_1)N^k` and so the velocity gets absorbed by the first term?